src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -814,51 +1179,150 @@ static inline ASR::expr_t* instantiate_Any(Allocator &al, const Location &loc,
+        Vec<ASR::expr_t *> e_args; e_args.reserve(al, 1);
+        e_args.push_back(al, s_var);
+        e_args.push_back(al, arg);
+
+        ASR::ttype_t *return_type = b.Tuple(loc, {character(-2), character(-2), character(-2)});
+        ASR::expr_t *value = nullptr;
+        if (ASR::is_a<ASR::StringConstant_t>(*s_var)
+         && ASR::is_a<ASR::StringConstant_t>(*arg)) {
+            ASR::StringConstant_t* sep = ASR::down_cast<ASR::StringConstant_t>(arg);
+            std::string s_sep = sep->m_s;
+            ASR::StringConstant_t* s_str = ASR::down_cast<ASR::StringConstant_t>(s_var);
+            std::string s_val = s_str->m_s;
+            if (s_sep.size() == 0) {
+                err("Separator cannot be an empty string", sep->base.base.loc);
+            }
+            value = eval_Partition(al, loc, s_val, s_sep);
+        }
+
+        return ASR::make_IntrinsicFunction_t(al, loc,
+            static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Partition),
+            e_args.p, e_args.n, 0, return_type, value);
+    }
+
+    static inline ASR::expr_t *instantiate_Partition(Allocator &al,
+        const Location &loc, SymbolTable *scope,
+        Vec<ASR::ttype_t*>& /*arg_types*/, Vec<ASR::call_arg_t>& new_args,
+        int64_t /*overload_id*/, ASR::expr_t* compile_time_value)
+    {
+        declare_function_variables("_lpython_str_partition");
+        {
+            args.reserve(al, 2);
+            Variable(target_string, character(-2), In);
+            args.push_back(al, target_string);
+            Variable(pattern, character(-2), In);
+            args.push_back(al, pattern);
+        }
+        ASR::ttype_t *return_type = b.Tuple(loc, {character(-2), character(-2), character(-2)});
+        Variable(result, return_type, ReturnVar)
+        {
+            Variable(index, int32, Local);
+            body.push_back(al, Assign(index, b.Call(UnaryIntrinsicFunction::
+                create_KMP_function(al, loc, scope), args, int32, loc)));
+            body.push_back(al, b.If(loc, iEq(index, i32_n(-1)), {
+                    Assign(result, b.TupleConstant(loc, {
+                        args[0], StringConstant("", character(0)),
+                        StringConstant("", character(0)) },
+                    b.Tuple(loc,{character(-2), character(0), character(0)})))
+                }, {
+                    Assign(result, b.TupleConstant(loc, {
+                        StringSection(args[0], i32(0), index), args[1],
+                        StringSection(args[0], iAdd(index, StringLen(args[1])),
+                            StringLen(args[0]))}, return_type))
+                }));
+            body.push_back(al, STMT(ASR::make_Return_t(al, loc)));
+        }
+        ASR::symbol_t *fn_sym = make_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, Source, Implementation, nullptr);
+        scope->add_symbol(fn_name, fn_sym);
+        return b.Call(fn_sym, new_args, return_type, compile_time_value, loc);
+    }
```

Comment:
I think this is the max we can simplify here.
What do you think about this?

@certik @czgdp1807

Reply:
I used this as a reference: https://github.com/lcompilers/lpython/pull/1673#discussion_r1172122992

Reply:
This is good. We can do better with changing `body` to just use `{}`, but let's do that later. For now this looks good.